internal/webhook/external/github.go

```diff
@@ -6,57 +6,66 @@ import (
 	"net/http"
 
 	gh "github.com/google/go-github/v71/github"
-	corev1 "k8s.io/api/core/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/git"
 	xhttp "github.com/akuity/kargo/internal/http"
 	"github.com/akuity/kargo/internal/io"
 	"github.com/akuity/kargo/internal/logging"
 )
 
-// githubHandler handles push events for github.
-// After the request has been authenticated,
-// the kubeclient is queried for all warehouses that contain a subscription
-// to the repo in question. Those warehouses are then patched with a special
-// annotation that signals down stream logic to refresh the warehouse.
-func githubHandler(
+// githubWebhookReceiver is an implementation of WebhookReceiver that handles
+// inbound webhooks from GitHub.
+type githubWebhookReceiver struct {
+	*baseWebhookReceiver
+}
+
+// newGitHubWebhookReceiver returns a new instance of githubWebhookReceiver.
+func newGitHubWebhookReceiver(
 	c client.Client,
-	namespace string,
-	secretName string,
-) http.HandlerFunc {
+	project string,
+	cfg kargoapi.WebhookReceiverConfig,
+) WebhookReceiver {
+	return &githubWebhookReceiver{
+		baseWebhookReceiver: &baseWebhookReceiver{
+			client:     c,
+			project:    project,
+			secretName: cfg.GitHub.SecretRef.Name,
+		},
+	}
+}
+
+// GetDetails implements WebhookReceiver.
+func (g *githubWebhookReceiver) getReceiverType() string {
+	return "github"
+}
+
+// getSecretValues implements WebhookReceiver.
+func (g *githubWebhookReceiver) getSecretValues(
+	secretData map[string][]byte,
+) ([]string, error) {
+	token, ok := secretData["token"]
```

Comment:
Should `token` be a `const`?

Reply:
It should. Will do.

Separately, I have reservations about it being called "token." When registering a webhook in GH, they call it "secret."

I'll have to double-check docs to see if they make frequent use of the word "token" in describing the registration process, but even if they did, my gut says matching what it's called on the registration form would remain more intuitive.

Any thoughts?

Reply:
They use "token" in their [titles](https://docs.github.com/en/webhooks/using-webhooks/validating-webhook-deliveries#creating-a-secret-token), but then `secret`/`payload` in the field names of their [example](https://docs.github.com/en/webhooks/using-webhooks/validating-webhook-deliveries#testing-the-webhook-payload-validation). If you look at the SDK, it is [called a `secretToken`](https://pkg.go.dev/github.com/google/go-github/v72/github#ValidateSignature) 🤷.

---

Unrelated, but worth noting now that I looked at the SDK: we may be interested in using [`ValidatePayload`](https://pkg.go.dev/github.com/google/go-github/v72/github#ValidatePayload).